### PR TITLE
Add preprocessed hazardsets

### PR DESCRIPTION
**Pull request by arnaud-morvan**
- Change decision tree by pure SQL one, less memory and time consuming.
- Process preprocessed hazardsets.
- Complete readme with detailed informations on settings.

Depends on https://github.com/GFDRR/thinkhazard_common/pull/10

Note that we don't force anymore the ZopeTransaction keep_session to True. We use only the hazardset id in main loop so we don't keep any object from the session after transaction commit or abort call.

Upgrade database with : 

```
ALTER TABLE processing.layer
    DROP CONSTRAINT layer_pkey,
    DROP CONSTRAINT layer_geonode_id_key;

ALTER TABLE processing.layer
    ADD CONSTRAINT layer_pkey PRIMARY KEY (geonode_id);

ALTER TABLE processing.layer
    ALTER COLUMN hazardlevel_id DROP NOT NULL,
    ALTER COLUMN return_period DROP NOT NULL,
    ALTER COLUMN hazardunit DROP NOT NULL;
ALTER TABLE processing.layer
    ADD CONSTRAINT layer_hazardlevel_id_key UNIQUE(hazardset_id, hazardlevel_id);
```

### CODE REVIEW

File: thinkhazard_processing/decisiontree.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    return '''
+INSERT INTO datamart.rel_hazardcategory_administrativedivision (
+    administrativedivision_id,
+    hazardcategory_id,
+    source
+)
+SELECT DISTINCT
+    o.admin_id AS administrativedivision_id,
+    first_value(hc.id) OVER w AS hazardcategory_id,
+    first_value(hs.id) OVER w AS source
```
Comment (fvanderbiest) [inline thinkhazard_processing/decisiontree.py]: From the thor mailing list: we should use the `owner.organization` field from the GeoNode API instead.

``` json
{
    "owner": {
        "organization": "WorldBank",
        "...": "..."
```

This will be done in another PR, after this one is merged.

**Comment by fvanderbiest** [comment]
I would have liked the decision tree rewrite (https://github.com/GFDRR/thinkhazard_processing/commit/00f27ea8f1289407f0b09d6c3cdc989a77a52d46) to be a standalone PR.

**Comment by fvanderbiest** [comment]
https://github.com/GFDRR/thinkhazard_processing/pull/26 being merged, I have rebased this one onto master.

Reviewing should be easier.

File: README.rst [hunk 2]
```diff
@@ -22,6 +22,106 @@ For more options, see::
...
+Database connection parameters, example :
+
+.. code:: yaml
+
+    sqlalchemy.url: postgresql://www-data:www-data@localhost:5432/thinkhazard_processing
+
+data_path
+---------
+
+Path to main data folder, example :
```
Comment (fvanderbiest) [inline README.rst]: Maybe mention that it should be a dedicated partition, to prevent side-effects if/when data is huge.

File: README.rst [hunk 3]
```diff
@@ -22,6 +22,106 @@ For more options, see::
...
+One entry for each hazard type mnemonic, with subkeys :
+
+- ``hazard_type`` : Corresponding hazard_type value in geonode.
+
+- ``preprocessed`` : Mandatory processing method. Possible values are :
+
+  1) ``True`` : One preprocessed layer with values corresponding directly to
+     hazard levels.
+
+     Need a corresponding ``values`` configuration.
```
Comment (fvanderbiest) [inline README.rst]: Needs

File: README.rst [hunk 4]
```diff
@@ -22,6 +22,106 @@ For more options, see::
...
+  .. code:: yaml
+
+      return_periods:
+        HIG: [10, 25]
+        MED: 50
+        LOW: [100, 1000]
+
+- ``thresholds`` : Flexible threshold configuration.
+
+  This can be a simple and global value for hazardtype. Example:
```
Comment (fvanderbiest) [inline README.rst]: per hazardtype ?

File: README.rst [hunk 5]
```diff
@@ -22,6 +22,106 @@ For more options, see::
...
+
+  .. code:: yaml
+
+       thresholds: 1700
+
+  But it can also contains one or many sublevels for complex configurations :
```
Comment (fvanderbiest) [inline README.rst]: contain

File: README.rst [hunk 6]
```diff
@@ -22,6 +22,106 @@ For more options, see::
...
+
+  1) ``global`` and ``local`` entries for corresponding hazardsets.
+  2) One entry for each hazard level mnemonic.
+  3) One entry for each hazard unit from geonode.
```
Comment (fvanderbiest) [inline README.rst]: /for each/per/

File: README.rst [hunk 7]
```diff
@@ -34,6 +134,21 @@ For example, you can define a specific database connection with a
 
     sqlalchemy.url: postgresql://www-data:www-data@localhost:9999/thinkhazard
 
+Processing tasks
+================
+
+Thinkhazard_processing provide some consecutive tasks to produce data for the
```
Comment (fvanderbiest) [inline README.rst]: provides

File: requirements.txt [hunk 8]
```diff
@@ -5,5 +5,5 @@ PyYAML==3.11
 rasterio==0.30.0
 pyproj==1.9.4
 shapely==1.5.13
-git+https://github.com/GFDRR/thinkhazard_common.git#egg=thinkhazard_common
+git+https://github.com/GFDRR/thinkhazard_common.git@fbbadb3f50722541c0cc701bd20c31c6c3bb7d7c#egg=thinkhazard_common
```
Comment (fvanderbiest) [inline requirements.txt:8]: won't be required since https://github.com/GFDRR/thinkhazard_common/pull/10 has been merged

File: thinkhazard_processing.yaml [hunk 9]
```diff
@@ -1,37 +1,78 @@
...
-      return_periods:
-        LOW: 1000
-        MED: 50
-        HIG: 25
-    local:
-      return_periods:
-        LOW: 100
-        MED: 50
-        HIG: 25
+    preprocessed: False
```
Comment (fvanderbiest) [inline thinkhazard_processing.yaml]: Since most hazard types are not preprocessed, this line should not be required IMO.

File: thinkhazard_processing/models.py [hunk 10]
```diff
@@ -77,26 +77,27 @@ def layerByLevel(self, level):
...
-                          primary_key=True)
+                          nullable=False)
     # the related hazard_level, inferred from return_period
     hazardlevel_id = Column(Integer,
-                            ForeignKey('datamart.enum_hazardlevel.id'),
-                            primary_key=True)
+                            ForeignKey('datamart.enum_hazardlevel.id'))
     # the return period is typically 100, 475, 2475 years but it can vary
-    return_period = Column(Integer, nullable=False)
+    return_period = Column(Integer)
```
Comment (fvanderbiest) [inline thinkhazard_processing/models.py:91]: could you explain why `return_period` and `hazardunit` should be nullable ?

File: thinkhazard_processing/processing.py [hunk 11]
```diff
@@ -29,35 +27,60 @@
...
     if hazardset_id is not None:
-        hazardsets = hazardsets.filter(HazardSet.id == hazardset_id)
+        ids = ids.filter(HazardSet.id == hazardset_id)
     if not force:
-        hazardsets = hazardsets.filter(HazardSet.processed.is_(False))
-    if hazardsets.count() == 0:
-        print 'No hazardsets to process'
+        ids = ids.filter(HazardSet.processed.is_(False))
+    if ids.count() == 0:
+        logger.info('No hazardsets to process')
```
Comment (fvanderbiest) [inline thinkhazard_processing/processing.py]: No hazardset to process

File: thinkhazard_processing/processing.py [hunk 12]
```diff
@@ -66,144 +89,283 @@ def process_hazardset(hazardset, force=False):
...
-            .one()
-        layers[level] = layer
+    type_settings = settings['hazard_types'][hazardset.hazardtype.mnemonic]
 
     with rasterio.drivers():
-        with rasterio.open(layers['HIG'].path()) as src_hig, \
-                rasterio.open(layers['MED'].path()) as src_med, \
-                rasterio.open(layers['LOW'].path()) as src_low:
+        try:
+            logger.info("  Openning raster files")
```
Comment (fvanderbiest) [inline thinkhazard_processing/processing.py]: Opening

File: thinkhazard_processing/processing.py [hunk 13]
```diff
@@ -66,144 +89,283 @@ def process_hazardset(hazardset, force=False):
...
-                            func.ST_GeomFromText(polygon.wkt, 4326))) \
-                    .filter(func.ST_Intersects(
-                        func.ST_Transform(AdministrativeDivision.geom, 4326),
-                        func.ST_GeomFromText(polygon.wkt, 4326)))
-
-            current = 0
-            outputs = 0
-            total = admindivs.count()
-            for admindiv in admindivs:
-                # print ' ', admindiv.id, admindiv.code, admindiv.name
```
Comment (fvanderbiest) [inline thinkhazard_processing/processing.py]: comment to be removed

File: thinkhazard_processing/processing.py [hunk 14]
```diff
@@ -66,144 +89,283 @@ def process_hazardset(hazardset, force=False):
...
+                hazardlevel = preprocessed_hazardlevel(hazardset,
+                                                       layers[0], readers[0],
+                                                       reprojected)
+            else:
+                hazardlevel = notpreprocessed_hazardlevel(hazardset,
+                                                          layers, readers,
+                                                          reprojected)
+
+        except Exception as e:
+            e.message = ("{}-{} raise an exception :\n{}"
```
Comment (fvanderbiest) [inline thinkhazard_processing/processing.py]: raises

File: thinkhazard_processing/scripts/process.py [hunk 15]
```diff
@@ -15,6 +15,10 @@ def main(argv=sys.argv):
         '--force', dest='force',
         action='store_const', const=True, default=False,
         help='Force execution even if hazardset has already been processed')
+    parser.add_argument(
+        '--dry-run', dest='dry_run',
+        action='store_const', const=True, default=False,
+        help='Perform a trial run that do not commit changes')
```
Comment (fvanderbiest) [inline thinkhazard_processing/scripts/process.py]: does not commit changes

File: requirements.txt (hunk 8 above)
Comment (pgiraud) [inline requirements.txt:8]: @fvanderbiest we decided to pin the versions of thinkhazard_common from now on. Helps preventing travis to fail.
For example, when someone needs a new version of thinkhazard_common, there's no need to wait for the modifications to be merged.

File: thinkhazard_processing/models.py (hunk 10 above)
Comment (fvanderbiest) [inline thinkhazard_processing/models.py:91]: Got it, pre-processed layers do not have a return period nor a unit.

File: README.rst [hunk 16]
```diff
@@ -22,6 +22,108 @@ For more options, see::
...
+
+hazard_types
+------------
+
+Harvesting and processing configuration for each hazard type.
+One entry for each hazard type mnemonic, with subkeys:
+
+- ``hazard_type``: Corresponding hazard_type value in geonode.
+
+- ``preprocessed``: Mandatory processing method. Possible values are:
```
Comment (fvanderbiest) [inline README.rst]: I don't see the use of `preprocessed`.
If the admin provides a `values` block, then we should consider the hazardtype as preprocessed.
If not, there should be `return_periods` and `thresholds`.

File: requirements.txt (hunk 8 above)
Comment (fvanderbiest) [inline requirements.txt:8]: OK, then we have to update it to latest rev.

File: README.rst (hunk 16 above)
Comment (fvanderbiest) [inline README.rst]: Fixed with e737ff9

**Comment by pgiraud** [comment]
@fvanderbiest Can you please merge?
